flink-connector-examples/src/main/java/io/pravega/example/flink/streamcuts/process/StreamBookmarker.java

```diff
@@ -0,0 +1,203 @@
+                                                              .addSink(writer);
+
+        // Execute within the Flink environment.
+        env.execute("StreamBookmarker");
+        LOG.info("Ending StreamBookmarker...");
+    }
+}
+
+/**
+ * This class processes the events read from the Pravega stream. Moreover, it also instantiates a reader group to check
+ * the state of the Flink reader. The main task of this class is to output SensorStreamSlice objects that represent events
+ * created by DataProducer whose value is < 0.
+ */
+class Bookmarker extends ProcessFunction<Tuple2<Integer, Double>, SensorStreamSlice>{
+
+    private static final Logger LOG = LoggerFactory.getLogger(Bookmarker.class);
+    private final URI pravegaControllerURI;
+
+    private transient ValueState<StreamCut> startStreamCut;
+    private transient ValueState<StreamCut> lastStreamCutBeforeSlice;
+    private transient ValueState<Tuple2<Integer, Double>> lastValue;
+
+    public Bookmarker(URI pravegaControllerURI) {
+        this.pravegaControllerURI = pravegaControllerURI;
+    }
+
+    @Override
+    public void open(Configuration parameters) {
+        // We will provide Bookmarker with a pointer to see the state of the reader group.
+        startStreamCut = getRuntimeContext().getState(new ValueStateDescriptor<>("startStreamCut",
+                TypeInformation.of(new TypeHint<StreamCut>() {})));
+        lastStreamCutBeforeSlice = getRuntimeContext().getState(new ValueStateDescriptor<>("lastStreamCutBeforeSlice",
+                TypeInformation.of(new TypeHint<StreamCut>() {})));
+        lastValue = getRuntimeContext().getState(new ValueStateDescriptor<>("lastValue",
+                TypeInformation.of(new TypeHint<Tuple2<Integer, Double>>() {})));
+    }
+
+    /**
+     * This function inspects the events written by a particular sensor (keyed stream by sensor id). When this function
+     * finds that there are events < 0 it considers them as of interest for further processing. For this reason, it
+     * gets a StreamCut related to the first observed event < 0 (start of stream slice). Then, if events become > 0
+     * again, the function schedules a timer to get a future StreamCut (end of slice). That is, the SensorStreamSlice
+     * object written to the collector will contain all the events < 0 for a certain sensor.
+     *
+     * @param value Input tuple from a sensor
+     * @param ctx Context
+     * @param out Collector
+     * @throws IOException
+     */
+    @Override
+    public void processElement(Tuple2<Integer, Double> value, Context ctx, Collector<SensorStreamSlice> out) throws IOException {
+        if (lastValue.value() != null) return;
+        // Infer the slice of the stream for which events are < 0.
+        if (value.f1 < 0 && startStreamCut.value() == null) {
+            startStreamCut.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
```

Comment:
Can we not initialize the reader group instance in `open()` and use it here?

Reply:
I found that problematic, so I optimized the way of getting the reader group by having a `ReaderGroup` attribute per task and initializing it only when it is `null`.